In [1]:
import requests
import json
import time

### 1. Obtendo o código dos municipios

In [2]:
url = "https://api-dados-abertos.tce.ce.gov.br/municipios"

headers = {
    "accept": "application/json"
}

In [3]:
response = requests.get(url, headers=headers)

In [4]:
response.json()

{'data': [{'codigo_municipio': '001',
   'nome_municipio': 'T.C.M.',
   'geoibgeId': None,
   'geonamesId': None},
  {'codigo_municipio': '002',
   'nome_municipio': 'ABAIARA',
   'geoibgeId': '2300101',
   'geonamesId': '6320034'},
  {'codigo_municipio': '003',
   'nome_municipio': 'ACARAPE',
   'geoibgeId': '2300150',
   'geonamesId': '6320035'},
  {'codigo_municipio': '004',
   'nome_municipio': 'ACARAU',
   'geoibgeId': '2300200',
   'geonamesId': '6320036'},
  {'codigo_municipio': '005',
   'nome_municipio': 'ACOPIARA',
   'geoibgeId': '2300309',
   'geonamesId': '6324372'},
  {'codigo_municipio': '006',
   'nome_municipio': 'AIUABA',
   'geoibgeId': '2300408',
   'geonamesId': '3408131'},
  {'codigo_municipio': '007',
   'nome_municipio': 'ALCANTARAS',
   'geoibgeId': '2300507',
   'geonamesId': '6320037'},
  {'codigo_municipio': '008',
   'nome_municipio': 'ALTANEIRA',
   'geoibgeId': '2300606',
   'geonamesId': '6320038'},
  {'codigo_municipio': '009',
   'nome_municipio': 'ALT

In [5]:
data = response.json()

In [6]:
print(data)

{'data': [{'codigo_municipio': '001', 'nome_municipio': 'T.C.M.', 'geoibgeId': None, 'geonamesId': None}, {'codigo_municipio': '002', 'nome_municipio': 'ABAIARA', 'geoibgeId': '2300101', 'geonamesId': '6320034'}, {'codigo_municipio': '003', 'nome_municipio': 'ACARAPE', 'geoibgeId': '2300150', 'geonamesId': '6320035'}, {'codigo_municipio': '004', 'nome_municipio': 'ACARAU', 'geoibgeId': '2300200', 'geonamesId': '6320036'}, {'codigo_municipio': '005', 'nome_municipio': 'ACOPIARA', 'geoibgeId': '2300309', 'geonamesId': '6324372'}, {'codigo_municipio': '006', 'nome_municipio': 'AIUABA', 'geoibgeId': '2300408', 'geonamesId': '3408131'}, {'codigo_municipio': '007', 'nome_municipio': 'ALCANTARAS', 'geoibgeId': '2300507', 'geonamesId': '6320037'}, {'codigo_municipio': '008', 'nome_municipio': 'ALTANEIRA', 'geoibgeId': '2300606', 'geonamesId': '6320038'}, {'codigo_municipio': '009', 'nome_municipio': 'ALTO SANTO', 'geoibgeId': '2300705', 'geonamesId': '3472474'}, {'codigo_municipio': '010', 'no

In [7]:
numero_municipios = len(data['data'])
numero_municipios

185

In [8]:
codigo_municipios = []
for municipio in data['data']:
    print(f"Código: {municipio['codigo_municipio']}, Nome: {municipio['nome_municipio']}")
    codigo_municipios.append(municipio['codigo_municipio'])

Código: 001, Nome: T.C.M.
Código: 002, Nome: ABAIARA
Código: 003, Nome: ACARAPE
Código: 004, Nome: ACARAU
Código: 005, Nome: ACOPIARA
Código: 006, Nome: AIUABA
Código: 007, Nome: ALCANTARAS
Código: 008, Nome: ALTANEIRA
Código: 009, Nome: ALTO SANTO
Código: 010, Nome: AMONTADA
Código: 011, Nome: ANTONINA DO NORTE
Código: 012, Nome: APUIARES
Código: 013, Nome: AQUIRAZ
Código: 014, Nome: ARACATI
Código: 015, Nome: ARACOIABA
Código: 016, Nome: ARARIPE
Código: 017, Nome: ARATUBA
Código: 018, Nome: ARNEIROZ
Código: 019, Nome: ASSARE
Código: 020, Nome: AURORA
Código: 021, Nome: BAIXIO
Código: 022, Nome: BANABUIU
Código: 023, Nome: BARBALHA
Código: 024, Nome: BARREIRA
Código: 025, Nome: BARRO
Código: 026, Nome: BARROQUINHA
Código: 027, Nome: BATURITE
Código: 028, Nome: BEBERIBE
Código: 029, Nome: BELA CRUZ
Código: 030, Nome: BOA VIAGEM
Código: 031, Nome: BREJO SANTO
Código: 032, Nome: CAMOCIM
Código: 033, Nome: CAMPOS SALES
Código: 034, Nome: CANINDE
Código: 035, Nome: CAPISTRANO
Código: 036, 

In [9]:
codigo_municipios

['001',
 '002',
 '003',
 '004',
 '005',
 '006',
 '007',
 '008',
 '009',
 '010',
 '011',
 '012',
 '013',
 '014',
 '015',
 '016',
 '017',
 '018',
 '019',
 '020',
 '021',
 '022',
 '023',
 '024',
 '025',
 '026',
 '027',
 '028',
 '029',
 '030',
 '031',
 '032',
 '033',
 '034',
 '035',
 '036',
 '037',
 '038',
 '039',
 '040',
 '041',
 '042',
 '043',
 '044',
 '045',
 '046',
 '047',
 '048',
 '049',
 '050',
 '051',
 '052',
 '053',
 '054',
 '055',
 '056',
 '057',
 '058',
 '059',
 '060',
 '061',
 '062',
 '063',
 '064',
 '065',
 '066',
 '067',
 '068',
 '069',
 '070',
 '071',
 '072',
 '073',
 '074',
 '075',
 '076',
 '077',
 '078',
 '079',
 '080',
 '081',
 '082',
 '083',
 '084',
 '085',
 '086',
 '087',
 '088',
 '089',
 '090',
 '091',
 '092',
 '093',
 '094',
 '095',
 '096',
 '097',
 '098',
 '099',
 '100',
 '101',
 '102',
 '103',
 '104',
 '105',
 '106',
 '107',
 '108',
 '109',
 '110',
 '111',
 '112',
 '113',
 '114',
 '115',
 '116',
 '117',
 '118',
 '119',
 '120',
 '121',
 '122',
 '123',
 '124',
 '125',


In [10]:
data_licitacao = ['2010-01-01_2010-12-31',
                 '2011-01-01_2011-12-31',
                 '2012-01-01_2012-12-31',
                 '2013-01-01_2013-12-31',
                 '2014-01-01_2014-12-31',
                 '2015-01-01_2015-12-31',
                 '2016-01-01_2016-12-31',
                 '2017-01-01_2017-12-31',
                 '2018-01-01_2018-12-31',
                 '2019-01-01_2019-12-31',
                 '2020-01-01_2020-12-31',
                 '2021-01-01_2021-12-31',
                 '2022-01-01_2022-12-31',
                 '2023-01-01_2023-12-31']

### 2. Obtendo  a base de dados dos licitantes de 185 municipios (2010 a 2023).

In [11]:
# https://api-dados-abertos.tce.ce.gov.br/licitacoes?codigo_municipio=057&data_realizacao_autuacao_licitacao=2010-01-01_2010-12-31
# url = "https://api-dados-abertos.tce.ce.gov.br/municipios"
# https://api-dados-abertos.tce.ce.gov.br/licitantes?codigo_municipio=057&data_realizacao_licitacao=2010-01-01_2010-12-31&quantidade=100&deslocamento=0
url1 = "https://api-dados-abertos.tce.ce.gov.br/licitantes?codigo_municipio="
url2 = "&data_realizacao_licitacao="
url3 = "&quantidade=100&deslocamento=0"
headers = {
    "accept": "application/json"
}


In [12]:
for i in codigo_municipios:
    for j in data_licitacao:
        sucesso = False
        retries = 0
        while not sucesso:
            response = requests.get(url1 + i + url2 + j + url3, headers=headers)
            if response.status_code == 200: # verifica se o servidor respondeu
                sucesso = True
                data = response.json()
                if len(data['data']) > 0:
                    with open(i + "_" + j + "_" + "licitante.json", "w") as json_file:
                        json.dump(data, json_file, indent=4)
            else:
                print("Erro " + i + "_" + j)
                time.sleep(10)
                retries += 1
                if retries >= 2: # realiza até 2 tentativas, em intervalos de 10 segundos
                    sucesso = True